### PR TITLE
KAFKA-5249: Fix incorrect producer snapshot offsets when recovering segments

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
@junrao Maybe you can review this?

**Comment by junrao** [comment]
@hachikuji : Thanks for the patch. LGTM.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 1]
```diff
@@ -2267,6 +2267,110 @@ class LogTest {
...
+    appendPid4(8) // 67
+    appendPid2(7) // 74
+    appendEndTxnMarkerAsLeader(log, pid2, epoch, ControlRecordType.ABORT) // 75
+    appendNonTransactionalAsLeader(log, 10) // 85
+    appendPid4(4) // 89
+    appendEndTxnMarkerAsLeader(log, pid4, epoch, ControlRecordType.COMMIT) // 90
+
+    // delete all snapshot files
+    logDir.listFiles.filter(f => f.isFile && f.getName.endsWith(Log.PidSnapshotFileSuffix)).foreach { file =>
+      file.delete()
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:2356]: Nit: we should probably use `Files.delete` to ensure an exception is thrown if the delete fails.

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/log/LogTest.scala:2356]: Ack. Maybe I can fix that in #3040?

Comment (ijuma) [reply core/src/test/scala/unit/kafka/log/LogTest.scala:2356]: Sounds good.

**Pull request closed without merging**